apps/emqx/src/emqx_cluster_rpc.erl

```diff
@@ -0,0 +1,322 @@
+            {keep_state, Node, [{reply, From, {error, Reason}}, catch_up_after(1)]}
+    end;
+
+handle_event(_EventType, _EventContent, _StateName, _Node) ->
+    {keep_state_and_data, [del_mfa_after(5 + rand:uniform(5))]}.
+
+terminate(_Reason, _StateName, _Node) ->
+    ok.
+
+code_change(_OldVsn, StateName, Node, _Extra) ->
+    {ok, StateName, Node}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+catch_up(Node) ->
+    case get_next_mfa(Node) of
+        {atomic, caught_up} -> {next_state, ?IDLE, Node, [del_mfa_after(5 * 60)]};
+        {atomic, {still_lagging, NextId, MFA}} ->
+            case apply_mfa(NextId, MFA) of
+                ok ->
+                    Fun = fun() ->
+                        Rec = #cluster_rpc_cursor{node = Node, tnx_id = NextId},
+                        mnesia:write(?CLUSTER_CURSOR, Rec, write)
+                          end,
+                    case transaction(Fun) of
+                        {atomic, ok} -> catch_up(Node);
+                        _ -> {next_state, ?CATCH_UP, Node, [catch_up_after(1)]}
+                    end;
+                _ -> {next_state, ?CATCH_UP, Node, [catch_up_after(1)]}
+            end;
+        {aborted, _Reason} -> {next_state, ?CATCH_UP, Node, [catch_up_after(1)]}
+    end.
+
+get_next_mfa(Node) ->
+    Fun =
+        fun() ->
+            case mnesia:read(?CLUSTER_CURSOR, Node, write) of
+                [] ->
+                    LatestId = get_latest_id(),
+                    Cursor = #cluster_rpc_cursor{node = Node, tnx_id = max(LatestId - 1, 0)},
+                    mnesia:write(?CLUSTER_CURSOR, Cursor, write),
+                    case mnesia:read(?CLUSTER_MFA, LatestId) of
+                        [] -> caught_up;
+                        [#cluster_rpc_mfa{mfa = MFA}] -> {still_lagging, LatestId, MFA}
+                    end;
+                [#cluster_rpc_cursor{tnx_id = DoneTnxId}] ->
+                    case mnesia:read(?CLUSTER_MFA, DoneTnxId + 1) of
+                        [] -> caught_up;
+                        [#cluster_rpc_mfa{mfa = MFA}] -> {still_lagging, DoneTnxId + 1, MFA}
+                    end
+            end
+        end,
+    transaction(Fun).
+
+do_catch_up(ToTnxId, Node) ->
+    case mnesia:wread({?CLUSTER_CURSOR, Node}) of
+        [] ->
+            Rec = #cluster_rpc_cursor{tnx_id = ToTnxId, node = Node},
+            mnesia:write(?CLUSTER_CURSOR, Rec, write),
+            caught_up;
+        [#cluster_rpc_cursor{tnx_id = DoneTnxId}] when ToTnxId =< DoneTnxId -> caught_up;
+        [Rec = #cluster_rpc_cursor{tnx_id = DoneTnxId}] ->
+            CurTnxId = DoneTnxId + 1,
+            [#cluster_rpc_mfa{mfa = MFA}] = mnesia:read(?CLUSTER_MFA, CurTnxId),
+            mnesia:write(?CLUSTER_CURSOR, Rec#cluster_rpc_cursor{tnx_id = CurTnxId}, write),
+            apply_mfa(CurTnxId, MFA)
```

Comment:
maybe move this `mnesia:write` call to an internal function.
e.g.
```
commit(TnxId) ->
    ok = mnesia:write(?CLUSTER_CURSOR, Rec#cluster_rpc_cursor{tnx_id = CurTnxId}, write).
```
and maybe call apply_mfa before commit, but check results like:
```
case apply_mfa(CurTnxId, MFA) of
    ok -> ok = commit(CurTnxId);
    {error, Reason} -> mnesia:abort(Reason)
end.
```
it's more straightforward to follow the code in this flow.
otherwise we'll have switch to the caller if we want to know how exceptions are handled